src/oled_display_node.cpp

```diff
@@ -190,118 +204,138 @@ static uint8_t sh1106_init_bytes[SH1106_INIT_BYTE_COUNT] = {
+
+        msgset[0].msgs = msgs;
+        msgset[0].nmsgs = 2;                     	  // number of transaction segments (write and read)
+
+        // The ioctl here will execute I2C transaction with kernel enforced lock
+        if (ioctl(fd, I2C_RDWR, &msgset) < 0) {
+            ROS_ERROR("Failed to get bus access to I2C device %s!  ERROR: %s", i2cDevFile, strerror(errno));
+            retCode = -2;
+            goto exitWithFileClose;
+        }
     }
-
-    if (ioctl(fd, I2C_SLAVE, address) < 0) {      // Set the port options and addr of the dev
-        ROS_ERROR("%s: Unable to get bus access to talk to slave\n", THIS_NODE_NAME);
-        close (fd);
-        return -3;
+    else {
+	// The ioctl here will address the I2C slave device making it ready for 1 or more other bytes
+    	if (ioctl(fd, I2C_SLAVE, slaveAddress) != 0) {    // Set the port options and addr of the dev
+      	  ROS_ERROR("Failed to get bus access to I2C device %s!  ERROR: %s", i2cDevFile, strerror(errno));
+      	  retCode = -3;
+          goto exitWithFileClose;
+    	}
+
+        bytesRead = read(fd, pBuffer, numBytes);
+        if (bytesRead != numBytes) {             	  // Verify that the number of bytes we requested were read
+          ROS_ERROR("Failed to read from I2C device %s!  ERROR: %s", i2cDevFile, strerror(errno));
+          retCode = -9;
+          goto exitWithFileClose;
+        }
     }
-    // we are now free to transmit to the I2C port 1 to latch 1
+    retCode = numBytes;
 
-    if ((write(fd, outBuf, numChars)) != numChars) {    // Write commands to the i2c port
-        ROS_ERROR("%s: Error writing proper byte count of %d bytes to i2c slave\n", THIS_NODE_NAME, numChars);
-        close (fd);
-        return -4;
-    }
-    close (fd);
+    exitWithFileClose:
+        close(fd);
 
-    return 0;
+    exitWithNoClose:
+
+    return retCode;
 }
 
 /*
- * @name                i2c_BufferRead
- * @brief               Read one or more bytes from an I2C based device
+ * @name                i2c_write
+ * @brief               Write one or more bytes to the I2C based device
  *
- * @param               i2cDevFile      Name of the I2C device
- * @param               i2c7bitAddr     7-bit I2C bus address
- * @param               pBuffer         User 8-bit buffer for data return
- * @param               chipRegAddr     Address register within chip. Use < 0 to suppress reg addr
- * @param               numBytes        Number of bytes to be read from the chip
+ * @param               dispCtx         Context for display that holds type and hardware interface info
+ * @param               outBuf          User 8-bit buffer for data input
+ * @param               numBytes        Number of bytes to be written to the chip
  *
- * @return              Returns number of bytes read where 0 or less implies some form of failure
+ * @return              Returns 0 for ok or less implies some form of failure
```

Comment:
This is a problem for our long term goal of reuse of these read and write calls.

One of the high wants is that the i2c_write routine is not specific to display node.  I failed to explain this well enough perhaps.   We do not want dispCtx to be required, that is specific to oled display.   That is why it was done with totally generic interface before.   

In ROS each node guarentees atomic operation within itself but we want i2c to be of use in different nodes and that is the reason for a lock.